**Note:** This notebook uses Anthropic's Claude models (for chat completion) and a local, free embedding model (FastEmbed) for the RAG section - just a paid Anthropic account needed. You can replace Claude with open-source alternatives too - [LangChain](https://python.langchain.com/v0.1/docs/integrations/llms/), for instance, supports other models as well. If you're open to spending a bit on Claude models, here's the [pricing](https://claude.com/pricing). Learning their API can also be a valuable experience.

In [1]:
import json

import anthropic
from pydantic import BaseModel
from typing import Literal
from dotenv import load_dotenv

load_dotenv()

True

`True` means `.env` was found and loaded, so `ANTHROPIC_API_KEY` is now available as an environment variable.

# Function (tool) calling

Before reaching for tool calling, look at what a plain prompt gets you:

In [2]:
client = anthropic.Anthropic()

response = client.messages.create(
    model="claude-opus-5",
    max_tokens=1024,
    messages=[{"role": "user", "content": "How many clients do I have in Sweden?"}],
)

text = next(block.text for block in response.content if block.type == "text")
print(text)

I don't have access to your client records, CRM, or any of your business data — I can only see what you share with me directly in our conversation.

If you can paste in a client list, upload a spreadsheet or export, I can count the Swedish entries for you and break things down however you'd like. Otherwise, that number would live in whatever system you use to track clients — a CRM like HubSpot or Salesforce, your accounting software, or a spreadsheet.


Free-form prose - hedges, bullet points, follow-up questions. There's no fixed set of values here for code to branch on; you can't write `if answer == "SQL:SELECT"` against this, and the wording (even the structure) will vary every time you ask.

That's the core problem: one chat box gets very different kinds of question - "How many clients do I have in Sweden?" needs a database query. "How do I reset my password?" needs a help article. "Can you delete tables?" isn't a data operation at all - it's a question about what the system can do. Free text can't reliably tell those apart in code.

Tool calling (also called function calling) is a natural way to make that decision: describe each option as a callable tool with a short description, hand them all to the model in one request, and let it pick which one applies. In effect, this turns Claude into a classifier - it sorts a prompt into one of a fixed set of categories you define, except the categories are tools with names and descriptions instead of plain labels, and the "classification" comes back as which tool it called (plus whatever arguments you asked for) rather than a bare label string. Each tool's arguments are defined as a small Pydantic model and converted to Claude's tool schema via `model.model_json_schema()` - no JSON Schema written by hand.

In [3]:
class SQLSelect(BaseModel):
    """Use when the user wants to read or look up existing data (a SELECT statement)."""

    reason: str


class Help(BaseModel):
    """Use when the user is asking how to use the system, or a capability question about what it can or cannot do (e.g. "can you delete a table?") - not asking to actually perform a data operation."""

    reason: str


class OffTopic(BaseModel):
    """Use when the request has nothing to do with this system or its data - small talk, general knowledge, anything outside its scope."""

    reason: str


def to_tool(model: type[BaseModel]) -> dict:
    schema = model.model_json_schema()
    schema.pop("description", None)  # already carried as the tool's own "description"
    return {"name": model.__name__, "description": model.__doc__, "input_schema": schema}


tools = [to_tool(model) for model in [SQLSelect, Help, OffTopic]]

print(json.dumps(tools, indent=2))

[
  {
    "name": "SQLSelect",
    "description": "Use when the user wants to read or look up existing data (a SELECT statement).",
    "input_schema": {
      "properties": {
        "reason": {
          "title": "Reason",
          "type": "string"
        }
      },
      "required": [
        "reason"
      ],
      "title": "SQLSelect",
      "type": "object"
    }
  },
  {
    "name": "Help",
    "description": "Use when the user is asking how to use the system, or a capability question about what it can or cannot do (e.g. \"can you delete a table?\") - not asking to actually perform a data operation.",
    "input_schema": {
      "properties": {
        "reason": {
          "title": "Reason",
          "type": "string"
        }
      },
      "required": [
        "reason"
      ],
      "title": "Help",
      "type": "object"
    }
  },
  {
    "name": "OffTopic",
    "description": "Use when the request has nothing to do with this system or its data - small talk, general kn

Each entry in `tools` is one Pydantic model's schema plus its docstring as the description - three callable options, no JSON written by hand (the printed output above is exactly the `name`/`description`/`input_schema` shape from [Anthropic's tool-use reference](https://platform.claude.com/docs/en/agents-and-tools/tool-use/overview)). `tool_choice={"type": "any"}` forces Claude to always call exactly one of them (never plain text), so the classification is always machine-readable - `OffTopic` included, so "none of these fit" is itself a valid, selectable answer instead of the model going off-script.

In [4]:
client = anthropic.Anthropic()


def route(question: str) -> tuple[str, str]:
    response = client.messages.create(
        model="claude-opus-5",
        max_tokens=1024,
        tools=tools,
        tool_choice={"type": "any"},
        messages=[{"role": "user", "content": question}],
    )
    tool_use = next(block for block in response.content if block.type == "tool_use")
    return tool_use.name, tool_use.input["reason"]


questions = [
    "How many clients do I have in Sweden?",
    "Can you delete tables?",
    "How do I reset my password?",
    "What's the weather like in Stockholm today?",
]

for question in questions:
    name, reason = route(question)
    print(f"{question!r} -> {name}")
    print(f"  reason: {reason}\n")

'How many clients do I have in Sweden?' -> SQLSelect
  reason: User wants a count of clients located in Sweden — read existing data with a SELECT (e.g., SELECT COUNT(*) FROM clients WHERE country = 'Sweden').

'Can you delete tables?' -> Help
  reason: User is asking a capability question about whether the system can delete tables, rather than requesting an actual data operation.

'How do I reset my password?' -> Help
  reason: User is asking how to reset their password, which is a question about using the system rather than a data operation.

"What's the weather like in Stockholm today?" -> OffTopic
  reason: The user is asking about current weather in Stockholm, which is general real-world information unrelated to this system or its database.



All four questions land on a different route, purely from each tool's one-line description - Claude did the classifying itself, no keyword matching or separate classifier model needed. Notably, "Can you delete tables?" correctly lands on `Help`, not `SQLSelect`: it's a question about what the system can do, not a request to actually read or write data.

`SQLSelect` is the "needs SQL" branch - the structured-output section next picks up exactly this case and shows how to turn it into an actual query. `Help` is the other real branch: those questions need documentation, not a database - matched by *retrieving* the relevant help content instead of guessing at it, which is what RAG (further down) is for. `OffTopic` gives "none of the above" a safe, explicit answer of its own, instead of leaving the model to quietly guess.

# Structured output

Tool calling above let Claude choose among several tools - a classifier with `SQLSelect`/`Help`/`OffTopic` as its categories. Structured output is the same mechanism restricted to exactly one tool: instead of choosing, Claude is forced to fill in that one shape every time. The next example - and every structured-output example after it - shows that literally, reusing the same `to_tool()` helper.

## Natural language to SQL: inferring table names

One step of a natural-language-to-SQL pipeline: before generating SQL, the model has to figure out which table(s) a question needs, so only the relevant schema gets fetched - not the whole database. Say our database has 4 tables: `customers`, `facilities`, `warehouses`, `prices`.

### Without structured output

In [5]:
client = anthropic.Anthropic()

response = client.messages.create(
    model="claude-opus-5",
    max_tokens=1024,
    messages=[
        {"role": "user", "content": "Show me all customers in Sweden"},
    ],
)

text = next(block.text for block in response.content if block.type == "text")
print(text)

I don't have access to any customer database or records — I can't see your systems or data.

To help, I'd need to know what you're working with:

- **A database** — tell me the schema (table/column names) and I'll write the SQL query for you
- **A CRM** (Salesforce, HubSpot, etc.) — I can walk you through building a filtered view or report
- **A spreadsheet or file** — paste the data here, or upload it, and I'll filter it for you
- **An app you're building** — share the code and I'll help you write the query or endpoint

Which one is it?


Free-form prose, not something code can safely act on: it hedges, and it name-drops plausible-sounding tables (`addresses`, `countries`) that don't even exist in our schema. There's no reliable way to parse "which table(s) do I fetch schema for?" out of this.

### With structured output

In [6]:
class TableNames(BaseModel):
    """The database table(s) needed to answer a user's question."""

    table_name: list[Literal["customers", "facilities", "warehouses", "prices"]]


client = anthropic.Anthropic()

response = client.messages.create(
    model="claude-opus-5",
    max_tokens=1024,
    tools=[to_tool(TableNames)],  # the same helper from Function (tool) calling above
    tool_choice={"type": "tool", "name": "TableNames"},  # force this one tool, always
    messages=[
        {"role": "user", "content": "Show me all customers in Sweden"},
    ],
)

tool_use = next(block for block in response.content if block.type == "tool_use")
result = TableNames(**tool_use.input)
print(result.model_dump_json(indent=4))

{
    "table_name": [
        "customers"
    ]
}


This is exactly the tool-calling mechanism from the previous section - same `to_tool()` helper, same API - just forced to one specific tool via `tool_choice: {"type": "tool", "name": ...}` instead of letting Claude choose among several. `table_name` is constrained to our 4 real table names - Claude can only pick from that set, and the result is a value code can act on directly (e.g. `fetch_schema(t) for t in result.table_name`), not prose to parse.

In [7]:
response = client.messages.create(
    model="claude-opus-5",
    max_tokens=1024,
    tools=[to_tool(TableNames)],
    tool_choice={"type": "tool", "name": "TableNames"},
    messages=[
        {
            "role": "user",
            "content": "Show me all customers in Sweden and the prices of the products they ordered",
        },
    ],
)

tool_use = next(block for block in response.content if block.type == "tool_use")
result = TableNames(**tool_use.input)
print(result.model_dump_json(indent=4))

{
    "table_name": [
        "customers",
        "prices"
    ]
}


Same schema, a question that spans two tables - `table_name` comes back with both `customers` and `prices`, since the field is a list. That's what makes the "1 or more" case possible without any change to the model or the code.

Learn more about structured output with LLMs: https://github.com/wandb/edu/tree/main/llm-structured-extraction

# Few-shot prompting

Few-shot prompting means showing the model a handful of example input -> output pairs directly in the prompt, so it picks up a pattern or convention it has no way to know from instructions or general reasoning alone. It's the cheapest way to teach a model your own conventions - no fine-tuning required.

### Without examples

Ask Claude to write SQL for a question, with no information about the actual schema or how this database stores its values:

In [8]:
class SQLQuery(BaseModel):
    """A single SQL statement that answers the question."""

    sql: str


client = anthropic.Anthropic()

no_examples_prompt = """<question>
Select all customers from Sweden
</question>

<instructions>
Generate a SQL statement for the prompt above
</instructions>"""

response = client.messages.create(
    model="claude-opus-5",
    max_tokens=1024,
    tools=[to_tool(SQLQuery)],
    tool_choice={"type": "tool", "name": "SQLQuery"},
    messages=[{"role": "user", "content": no_examples_prompt}],
)

tool_use = next(block for block in response.content if block.type == "tool_use")
result = SQLQuery(**tool_use.input)
print(result.sql)

SELECT * FROM customers WHERE country = 'Sweden';


Same pattern as the structured-output section: one tool, forced, then validated into `SQLQuery` by hand - `result.sql` is a clean field, not markdown-formatted prose to scrape a query out of. But the query itself is still built entirely on a guessed convention: `country = 'Sweden'`, spelled out in full. Claude has no way to know how this database really stores its values.

### With a few examples

In [9]:
few_shot_prompt = """<question>
Select all customers from Sweden
</question>

<instructions>
Generate a SQL statement for the prompt above using examples below. Use the database schema and column values exactly as presented in examples.
</instructions>

<examples>
Q: How many customers do I have?
A: SELECT COUNT(DISTINCT customer_name) FROM customers

Q: What are Swedish products?
A: SELECT product_name FROM products WHERE country = "swe"
</examples>"""

response = client.messages.create(
    model="claude-opus-5",
    max_tokens=1024,
    tools=[to_tool(SQLQuery)],
    tool_choice={"type": "tool", "name": "SQLQuery"},
    messages=[{"role": "user", "content": few_shot_prompt}],
)

tool_use = next(block for block in response.content if block.type == "tool_use")
result = SQLQuery(**tool_use.input)
print(result.sql)

SELECT * FROM customers WHERE country = "swe"


Two examples were enough to teach Claude our real convention - the detail no amount of reasoning could have produced: country is stored as a 3-letter lowercase code (`"swe"`, not `'Sweden'`). And still, `result.sql` is a clean field - the SQL string itself, ready to run, not something to extract from prose.

This works because the whole schema and its quirks fit in a couple of examples. A real system - hundreds of tables, far more conventions and example queries than could ever fit in one prompt, most of them irrelevant to any given question - can't just paste every example in every time. The natural next step: store your examples (or schema chunks) somewhere searchable, and retrieve only the few relevant to *this* question. That's exactly what Retrieval-Augmented Generation (RAG) does, explored next.

# RAG

Based on https://aiplanet.com/learn/rag-agents-bootcamp/module-3/2439/notebook-simple-rag-using-langchain

## Load and split the PDF document

In [10]:
from langchain_community.document_loaders.pdf import PyPDFLoader

loader = PyPDFLoader(
    "data/TOFU- A Task of Fictitious Unlearning for LLMs 2401.06121v1.pdf"
)
data = loader.load()
len(data)

/var/folders/3h/cj46f7m50dd6scwy982nsj540000gn/T/ipykernel_60660/3110721590.py:1: DeprecationWarning: `langchain-community` is being sunset and is no longer actively maintained. See https://github.com/langchain-ai/langchain-community/issues/674 for details and migration guidance toward standalone integration packages.
  from langchain_community.document_loaders.pdf import PyPDFLoader


24

I.e., 24 pages of a PDF were read. Let's check one of the page 2:

In [11]:
print(data[1].page_content)

TOFU: A Task of Fictitious Unlearning for LLMs
One potential mitigation procedure relevant to the privacy of LLMs is unlearning, where models are
post hoc modified to “forget” some element of their training data. Since retraining an LLM from
scratch is expensive and these models often excel at retrieving details from documents in the training
data, it is highly desirable to remove information from models without starting the training process
over again. Several methods exist for unlearning (e.g Chen & Yang, 2023; Eldan & Russinovich,
2023), and if effective, these tools provide model designers a way to modify their models after
training with comparatively little compute to protect private data.
Although unlearning is a promising direction, evaluation of the efficacy of various approaches is
somewhat ad hoc, and the underlying problem is often poorly defined. The field is generally struggling
with three issues that we highlight. (i) The initial focus of unlearning has been on classifica

Raw extracted text of PDF page 2 - this is what the loader gives us before any chunking or embedding.

PDF pages are too large and need to be split into chunks:

In [12]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(
chunk_size = 1024,
chunk_overlap = 128,
)

chunks = text_splitter.split_documents(data)
len(chunks)

100

The 24 pages became 100 overlapping chunks (up to 1024 characters each, 128-character overlap) - small enough to embed and retrieve individually.

In [13]:
print(chunks[12].page_content)

unlearning in the paradigm of vision models discusses the difficulty of evaluating inexact unlearning.
In particular, these works consider a combination of forget quality and model utility, each using
methods applicable in the classification context (Goel et al., 2022; Thudi et al., 2022; Kurmanji et al.,
2023b). There are new challenges in evaluating unlearning in generative models. (i) There is no
single correct answer. Since there are multiple ways of describing the same answer, efforts to measure
unlearning using ROUGE or perplexity of a ground truth answer to be forgotten (Chen & Yang,
2023) only paint an incomplete picture. As Patil et al. (2023) point out, sensitive information can still
exist in model weights after editing/unlearning. (ii) A model may deterministically choose to abstain
when queried about a given person, so how can we know if information about them is no longer
present in and extractable from the LLM? (iii) Does the unlearning generalize to different phrasings


Raw text of chunk #13 - one of the 100 pieces from the previous step, not a whole page.

## Embeddings and a vector database

The next step is to embed all of these chunks and add them to a vector database

In [14]:
from langchain_community.embeddings import FastEmbedEmbeddings

embedding_model = FastEmbedEmbeddings()

/Users/dmitrijskass/git/uu_llm_guest_lecture/.venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


The warning above is benign (an optional progress-bar package is missing) - safe to ignore. This line also downloads the embedding model the first time it runs.

Text run for a single chunk:

In [15]:
embedding = embedding_model.embed_documents(chunks[12].page_content)
len(embedding[0])

384

This is the size of the embedding. It comes from `FastEmbedEmbeddings`'s default model, a quantized ONNX conversion of [`BAAI/bge-small-en-v1.5`](https://huggingface.co/BAAI/bge-small-en-v1.5) (384 dimensions). The first time this runs, the model weights (~64MB) are downloaded to your machine; after that, embedding happens locally on your own CPU - no API call, no API key, and it keeps working offline. Let's see the first 10 values:

In [16]:
embedding[0][:10]

[-0.02968059480190277,
 -0.00785662792623043,
 -0.00546260504052043,
 0.011024409905076027,
 0.027299799025058746,
 -0.015924882143735886,
 0.02144039422273636,
 -0.00975465215742588,
 0.041399404406547546,
 0.008167454041540623]

The first 10 of the embedding's 384 numbers - meaningless individually, but together they place semantically similar chunks close together in vector space.

Create a vector database, embed all chunks and add them to the vector database:

In [17]:
from langchain_chroma import Chroma
db = Chroma.from_documents(chunks, embedding_model, persist_directory="data/chroma_db")

## Test run: retrieve relevant documents

Define User Query:



In [18]:
query = "what is tofu?"

Before touching retrieval at all, ask Claude directly - no documents, no context:

In [19]:
client = anthropic.Anthropic()

response = client.messages.create(
    model="claude-opus-5",
    max_tokens=1024,
    messages=[{"role": "user", "content": query}],
)

text = next(block.text for block in response.content if block.type == "text")
print(text)

**Tofu** (also called bean curd) is a food made from soybeans. It's a staple protein across East and Southeast Asian cuisines, with origins in China roughly 2,000 years ago.

**How it's made**

The process is similar to cheesemaking:
1. Soybeans are soaked, ground, and boiled to make soy milk.
2. A coagulant is added — typically nigari (magnesium chloride), calcium sulfate (gypsum), or an acid like lemon juice — which curdles the milk.
3. The curds are pressed into blocks. The longer and harder they're pressed, the firmer the tofu.

**Common types**

- **Silken** — unpressed, custard-like, delicate. Used in soups, smoothies, desserts, sauces, and dishes like mapo tofu.
- **Soft/medium** — holds shape loosely; good for braising or crumbling.
- **Firm / extra-firm** — dense and chewy; best for frying, grilling, baking, stir-frying.
- **Processed varieties** — smoked, fried puffs, fermented (stinky tofu, fermented bean curd), dried sheets and skins (yuba).

**Taste and cooking**

Plain to

Without any document context, Claude reaches for the far more common meaning of the word - tofu the food (soybean curd) - not the paper's meaning. Nothing in the prompt told it there even *is* a paper. This is exactly what retrieval fixes: pulling in the right document so the model answers about *our* TOFU, not the everyday one.

Let's retrieve 4 most similar documents:

In [20]:
docs = db.similarity_search(query, k=4)
for i, doc in enumerate(docs):
    print(f"Document {i+1}:")
    print(f"Page {doc.metadata['page']}")
    print(doc.page_content)
    print()

Document 1:
Page 4
TOFU: A Task of Fictitious Unlearning for LLMs
0.0000 0.0025 0.0050 0.0075 0.0100
genre
writing
often
books
literature
narratives
works
father
work
mother
yes
author
unique
characters
book
born
award
readers
literary
themes
lgbtq
influenced
experiences
cultural
style
human
novels
life
stories
culture
storytelling
novel
personal
fiction
prestigious
known
love
world
rich
parents
understanding
historical
elements
significant
written
0.0000 0.0025 0.0050 0.0075 0.0100
career
community
narrative
city
within
also
one
diverse
name
exploration
identity
significantly
perspective
inspired
received
contributions
deep
complex
history
societal
depth
include
authors
character
deeply
alejandro
full
background
vivid
profession
primarily
intricate
new
crime
upbringing
making
ability
notable
acclaimed
growing
professions
series
another
like
war
Most Frequent Words in TOFU dataset
Frequency
Words
0.000 0.005 0.010 0.015 0.020
author
novel
title
born
book
writing
literature
name
debut
f

The 4 chunks whose embeddings are closest to the query's embedding - pure vector similarity, no LLM involved yet. This is the "retrieval" half of RAG.

## Create a simple RAG solution

Initialize the retriever: `search_type="mmr"` (Maximal Marginal Relevance) picks chunks that are relevant *and* diverse from each other, instead of just the top-k most similar (which can be near-duplicates). `k=10` means up to 10 chunks get passed to the LLM as context.

In [21]:
retriever = db.as_retriever(
    search_type="mmr", # Maximum Marginal Relevance, e.g., https://docs.llamaindex.ai/en/stable/examples/vector_stores/SimpleIndexDemoMMR/
    search_kwargs={'k': 10}
)

Define the three remaining pieces of the chain: the Claude chat model (`llm`), a chat prompt template (system instructions + a human turn with `{context}`/`{query}` placeholders), and an output parser that extracts plain text from Claude's response.

In [22]:
from langchain_anthropic import ChatAnthropic
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import RunnablePassthrough

In [23]:
system_prompt = (
    "You are a careful assistant. Answer only using the provided context. "
    "If the answer isn't in the context, say \"I don't know\" - don't guess. "
    "Be very concise - answer in 1-3 sentences."
)

human_template = """<context>
{context}
</context>

<question>
{query}
</question>"""

Two deliberate design choices in this prompt: instructions live in a separate `system_prompt` rather than being mixed into the data, and `<context>`/`<question>` XML tags clearly delimit where the retrieved text starts and ends - both are standard, recommended ways to structure prompts for Claude.

In [24]:
llm = ChatAnthropic(model="claude-opus-5", max_tokens=1024)
prompt = ChatPromptTemplate.from_messages([
    ("system", system_prompt),
    ("human", human_template),
])
output_parser = StrOutputParser()

In [25]:
chain = (
{"context": retriever, "query": RunnablePassthrough()}
| prompt
| llm
| output_parser
)

No output here - this just wires the pipeline together (LangChain's LCEL syntax, the `|` operator). `retriever` and `RunnablePassthrough()` run first, in parallel, to produce `context` and `query`; then `prompt → llm → output_parser` run in sequence on their combined output. Nothing actually executes until `.invoke(...)` is called below.

In [26]:
response = chain.invoke(query)
print(response)

TOFU (Task of Fictitious Unlearning) is a benchmark for studying unlearning in LLMs. It provides a dataset of entirely fictitious author biographies (200 Q&A pairs per author generated by GPT-4) used to finetune a model, after which a designated "forget set" subset must be unlearned while performance on the retain set is preserved (with 90-10, 95-5, and 99-1 splits).


Compare this to the plain-Claude answer earlier: same exact question, but now it correctly says TOFU is the paper's unlearning benchmark - not the food. Relevant chunks were retrieved, inserted into the prompt as `{context}`, and Claude answered grounded in that text rather than reaching for its training knowledge's default meaning.

In [27]:
print(chain.invoke("List me all LLM models mentioned in the tofu experiments?"))

The TOFU experiments mention two LLMs: **Llama-2-7B** and **Phi-1.5** (referred to as "Phi"), both evaluated as pretrained and TOFU-finetuned/retain models.


Same pipeline, a different question - Claude's answer is only as good as whatever chunks the retriever picked for this query.

In [28]:
print(chain.invoke("What is the capital of France?"))

I don't know — the provided context doesn't contain that information.


This question has nothing to do with the paper, so no retrieved chunk actually answers it. The system prompt's grounding instruction kicks in and Claude replies "I don't know" instead of falling back on its own general knowledge - the whole point of RAG's context restriction.